### PR TITLE
🏞️ Implement streams commands 

### LINKED ISSUE #36: Stream API
- https://redis.io/commands#stream

### PR DESCRIPTION
**Pull request by Terkwood**
# Add stream commands

This change set resolves #36.   :ship: 

## API design :art: 

We gladly welcome feedback on the design.  You can read the method defs in `command.ts` to see the most up-to-date version of the command inputs & outputs

### Strongly Typed Message IDs
  
We implemented strong types for stream entry IDs:

```typescript
export interface XId {
  unixMs: number;
  seqNo: number;
}
```

We made an effort to ensure flexibility in the command interfaces. For instance, in `xadd`, you may also pass simple numbers, or paired numbers `[1000,10]`, or the special string `"*"`.  

Methods which _return_ `XId`s will always return the exact form, so that you can deconstruct them with minimal pain.

See `XId` and the related types in `streams.ts` for details.  

### Return type for XREAD 📕 

We offer an `interface` for the XREAD reply type which is hopefully less annoying to use than the Redis default array type.  Please note that we return field_value pairs as a `Map`.

```typescript
export interface XMessage {
  id: XId;
  field_values: Map<string, string>;
}
export type XReadStream = { key: string; messages: XMessage[] };
export type XReadReply = XReadStream[];
```

- [x] implement a nice return type for XREAD and XREADGROUP

### Allow input of record objects in XADD

In `xadd`, we chose to allow input of arbitrary `Record<string | number,string | number>` objects, as well as `Map<string | number,string | number>`. 

⚠ This allows greater flexibility for input, but may end up being a little confusing since the field_value outputs of `xread` and `xreadgroup` are always `Map<string,string>`.  ⚠

- [x] support record objects in XADD
- [x] Remove support for array of string field_values in XADD, touch up docs to reflect 
- [x] expose a single xadd implementation & interface that allows default options

## Commands to implement and test 

- [x] xadd
- [x] xread
- [x] xreadgroup 
- [x] xack
- [x] xclaim
- [x] xpending
- [x] xtrim
- [x] xdel
- [x] xrange
- [x] xrevrange
- [x] xinfo
- [x] xlen
- [x] xgroup

## docs and sanity checks

We made an effort to verify that argument names match the canonical names exposed by Redis.  One exception is references to `ID` in Redis docs, which we call `xid: XId` in order to match the parsed, typed flavor of data.  Minimal command form should be included in each doc string.  

The JsDoc approach isn't very nice since [only typescript export statements can trigger deno doc](https://github.com/denoland/deno/pull/4500), but at least these docs are available to developers via their editors. 

- [x] xack
- [x] xadd
- [x] xread
- [x] xreadgroup 
- [x] xclaim
- [x] xpending
- [x] xtrim
- [x] xdel
- [x] xrange
- [x] xrevrange
- [x] xinfo
- [x] xlen
- [x] xgroup

## special reply types to model 

Some hints for reply types https://github.com/Terkwood/redis-rs/blob/0d126f54d6970b931594bf677b09a368af70617d/src/streams.rs#L231

- [x] pending: use a [discriminated union](https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions) to represent the [various response formats](https://redis.io/commands/xpending) 
- [x] info: similar to above, and be careful to heed the disclaimers [in the Redis docs](https://redis.io/commands/xinfo) 

## Additional requirements

- [x] Update README with example
- [x] Test suite record cleanup
- [x] add test demonstrating multiple consumers handling unique messages (fine grained work splitting) https://github.com/keroxp/deno-redis/issues/79#issuecomment-633261275 and https://github.com/keroxp/deno-redis/issues/79#issuecomment-632824135
- [x] address every TODO
- [x] match doc params to names (xid vs id, etc) 
- [x] check that `isNumber` areas are sane 

### Always Remember

- pass `deno lint`  (see #98 for inspiration).  The CI will enforce this anyway.
- pass `deno fmt --check`.  The CI will enforce this also.

### CODE REVIEW

**Comment by Terkwood** [comment]
Some hints on reply types https://github.com/Terkwood/redis-rs/blob/0d126f54d6970b931594bf677b09a368af70617d/src/streams.rs#L231

**Comment by Terkwood** [comment]
This discussion around object(`Record`) vs `Map` in JS is fascinating. 

https://stackoverflow.com/questions/18541940/map-vs-object-in-javascript

It's difficult to infer that one is easily better than the other, in terms of choosing which to support for `xadd` and `xread`. 

So I'll move ahead with providing commands that work with both type, with the default being object /Record.

**Comment by Terkwood** [comment]
OK, that was harder than I thought.  Sticking with `Map` return type for XRead and XReadGroup

**Comment by Terkwood** [comment]
Abandoned this idea, and trimmed it out of the description to simplify.  Leaving it in the comments as a pack-rat, for now (might delete it later):


Oak framework has a nice facility for allowing the return of _either_ record objects or Maps. Let's support that both for XADD and XREAD, XREADGROUP. 

https://github.com/oakserver/oak/blob/f293b6f8df1b2fbf27ae0c69424e05fbda887521/helpers.ts#L28

https://github.com/oakserver/oak#getqueryctx-options

It looks nice in theory, but we had some difficulty figuring out how to mangle the return type for `xread` and `xreadgroup` to allow a _choice_ between `Record<string,string>` and `Map<string,string>`.  So for now, `xread` and `xreadgroup` simply return a `Map<string,string>`.

**Comment by Terkwood** [comment]
Apologies for the build failure spam. I'll reduce the granularity of my commits until I get my tests passing!

**Comment by Terkwood** [comment]
Just a couple of small areas left to fill out.  This will be out of draft status soon.

File: redis.ts [hunk 1]
```diff
@@ -1202,6 +1231,363 @@ class RedisImpl implements RedisCommands {
...
+    key: string,
+    group: string,
+  ) {
+    return this.execArrayReply<Raw>("XPENDING", key, group)
+      .then((raw) => {
+        // When XPENDING is called with just a key name
+        // and a consumer group name, it just outputs a
+        // summary about the pending messages in a given
+        // consumer group.
+        // This will return XPendingData or XPendingEmpty
```
Comment (Terkwood) [inline redis.ts:1449]: Remove

**Comment by uki00a** [comment]
Thanks @Terkwood!  I'll review this PR in the coming days.

File: tests/stream_test.ts [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+
+    // minimum options
+    const claimingConsumer = "responsible-process";
+    const firstClaimed = await client.xclaim(
+      key,
+      { group, consumer: claimingConsumer, minIdleTime },
+      1000,
+      2000,
+    );
+    switch (firstClaimed.kind) {
```
Comment (uki00a) [inline tests/stream_test.ts:540]: Can this switch statement be replaced with the following code?
```suggestion
    assert(firstClaimed.kind === "messages");
    assertEquals(firstClaimed.messages.length, 2);
    assertEquals(
      firstClaimed.messages[0].field_values,
      new Map(Object.entries({ "field": "foo" })),
    );
...
```

File: tests/stream_test.ts [hunk 3]
```diff
@@ -0,0 +1,810 @@
...
+    // the output for justIDs will have a different shape
+    const secondClaimedXIds = await client.xclaim(
+      key,
+      { group, consumer: claimingConsumer, minIdleTime, justXId: true },
+      [3000, 0],
+      [3000, 1],
+      [3000, 2],
+    );
+
+    switch (secondClaimedXIds.kind) {
```
Comment (uki00a) [inline tests/stream_test.ts:602]: Can this switch statement be replaced with the following code?
```suggestion
    assert(firstClaimed.kind === "justxid");
    assertEquals(
      secondClaimedXIds.xids,
      [
        { unixMs: 3000, seqNo: 0 },
        { unixMs: 3000, seqNo: 1 },
        { unixMs: 3000, seqNo: 2 },
      ],
    );
...
```

File: tests/stream_test.ts [hunk 4]
```diff
@@ -0,0 +1,810 @@
...
+        group,
+        consumer: claimingConsumer,
+        minIdleTime,
+        retryCount: 6,
+        force: true,
+      },
+      4000,
+      5000,
+    );
+    switch (thirdClaimed.kind) {
```
Comment (uki00a) [inline tests/stream_test.ts:680]: Can this switch statement be replaced with the following code?

```suggestion
    assert(thirdClaimed.kind === "messages");
    assertEquals(thirdClaimed.messages.length, 2);
    ...
```

File: tests/stream_test.ts [hunk 5]
```diff
@@ -0,0 +1,810 @@
...
+    // New consumer read one message with ">"
+    assertEquals(newConsumer.pending, 1);
+    // Old consumer read two messages with ">"
+    assertEquals(oldConsumer.pending, 2);
+
+    assertEquals(await client.xgroup_destroy(key, "newgroup"), 1);
+  });
+});
+
+const sleepMillis = (ms: number) => {
```
Comment (uki00a) [inline tests/stream_test.ts:808]: Please use `defer` (defined in `std/async/mod.ts`) instead.

```suggestion
import { defer as sleepMillis } from "../vendor/https/deno.land/std/async/mod.ts";
```

File: command.ts [hunk 6]
```diff
@@ -286,6 +313,352 @@ export type RedisCommands = {
   srem(key: string, ...members: string[]): Promise<Integer>;
   sunion(...keys: string[]): Promise<BulkString[]>;
   sunionstore(destination: string, ...keys: string[]): Promise<Integer>;
+  // Stream
+  /**
```
Comment (uki00a) [inline command.ts:317]: nice documents!

File: stream.ts [hunk 7]
```diff
@@ -0,0 +1,401 @@
+import { ConditionalArray, Raw } from "./command.ts";
+
+export interface XId {
+  unixMs: number;
+  seqNo: number;
+}
+
+export interface XMessage {
+  xid: XId;
+  field_values: Map<string, string>;
```
Comment (uki00a) [inline stream.ts:10]: Hmmm, it's hard to decide whether `Map` or `Record` is better... :thinking:

**Comment by uki00a** [comment]
Looks good to me overall, but a few things need to be fixed.
Please check it out.

**Comment by Terkwood** [comment]
Thanks for the fast review! I will be able to respond on Thursday or Friday! The comments look straightforward.

I don't have a strong opinion on Record vs Map for the return type from the xread and xreadgroup methods, but my ultimate rationale was that iterating thru Maps seems a bit less clunky: https://stackoverflow.com/questions/16174182/typescript-looping-through-a-dictionary

But for accessing individual fields, Record (dictionary `a.b`) syntax seems a little friendlier to the library user.

I'm happy to use either choice.

File: tests/stream_test.ts [hunk 8]
```diff
@@ -0,0 +1,810 @@
...
+      const data = await client.xreadgroup(
+        [{ key, xid }],
+        { group, consumer },
+      );
+
+      assertEquals(data[0].messages.length, 1);
+
+      assertEquals(data[0].messages[0].field_values.get("target"), payload);
+
+      await cleanupStream(client, key);
```
[the remark this answers is not in the document]
Comment (Terkwood) [reply tests/stream_test.ts:423]: Wow, great spot!  Thank you!

File: tests/stream_test.ts (hunk 3 above)
Comment (Terkwood) [reply tests/stream_test.ts:602]: Yes, this improves it !

File: tests/stream_test.ts [hunk 9]
```diff
@@ -0,0 +1,810 @@
...
+});
+
+test("xgroup create and destroy", async () => {
+  const groupName = "test-group";
+
+  const key = randomStream();
+
+  let created = await client.xgroup_create(key, groupName, "$", true);
+  assertEquals(created, "OK");
+  try {
```
[the remark this answers is not in the document]
Comment (Terkwood) [reply tests/stream_test.ts]: Altered, but for now the `Error` type and exact error message are left `undefined`.  The message text contains a bunch of whitespace and a newline and is a little tricky to match.  Haven't tracked down the Error class yet.  I can come back to this if it's a sticking point...

[the remark this answers is not in the document]
Comment (Terkwood) [reply tests/stream_test.ts]:
```text

[the remark this answers is not in the document]
Comment (Terkwood) [reply tests/stream_test.ts]: Aha, got the message part matching, at least!

File: tests/stream_test.ts [hunk 10]
```diff
@@ -0,0 +1,792 @@
...
+import { makeTest } from "./test_util.ts";
+import { parseXId } from "../stream.ts";
+import {
+  assertEquals,
+  assert,
+  assertNotEquals,
+  assertThrowsAsync,
+} from "../vendor/https/deno.land/std/testing/asserts.ts";
+import { delay } from "../vendor/https/deno.land/std/async/mod.ts";
+
```
[the remark this answers is not in the document]
Comment (Terkwood) [reply tests/stream_test.ts:11]: Oh, nice. Thanks, that one was tough to find! Will be back at keys some time today...

**Comment by uki00a** [comment]
@Terkwood 
IMO it's not a bad idea to adopt `Map` over `Record` (because `Map` keeps the order of the fields in the stream entry).
Having said that, I'd like to hear the opinions of other people, so I'll open the issue about it.

Let's merge this PR for now!

**Comment by uki00a** [comment]
Merged, thanks a lot @Terkwood!

**Comment by Terkwood** [comment]
Woohoo! Thanks, my pleasure!

**Comment by uki00a** [comment]
Released in [v0.11.0](https://github.com/denolib/deno-redis/releases/tag/v0.11.0)!

**Comment by Terkwood** [comment]
So cool!! Thanks!

**Pull request merged**